### DISCUSSION TITLE
What is TwinBasic and what's the future plans.

### DISCUSSION THREAD

**Discussion by vbBaka** [twinbasic/twinbasic#617] [General]
@WaynePhillipsEA You are needed here because I don't understand something and I want clarity here.
It seems that Im alone here, everything I say its going against.
My understanding was the TwinBasic was the Continuation of VB6. it was born to create a modern version of VB6, and always have the fully compatibility in mind.

But, of what I can understand, people here seems to dislike VB6 and want, if they could, change "a lot" in favor for other syntaxes and methods.

So what is the real purpose of TB?
is it just to accommodate VB6 people, so they can "migrate" their project and when that is done its time to change?

the spirit of keeping the VB6 "easiness" seems to be ignored.
the response is "we need to evolve", and "all people want".
like we are all experts. but the whole reason of Basic, is that to accommodate beginners and people that will never get to that expert level.

and its not about one operator, that is nothing in comparison of what I can get from all the members here.
theres no love to VB6 at all, just, old and wrongly made and now its time to fix it.

so, thats the plan? to fix VB6 and turn it into a strange basic that people on expert level will enjoy using?

**Comment by Kr00l** [comment]
I don't understand why you make such a storm.
tB will be and _ever_ be VB6 compatible.
Thus when you desire you can code like beforw in tB like in VB6 by ignoring the new features.

So why are you afraid? Many here love VB6 btw..

**Reply by vbBaka** [reply]
its mostly from what few members wrote:
- Add to twinbasic to facilitate migration into twinbasic but mark as deprecated
- We don't have the luxury of a clean slate so we must come up with ways to help converging toward a consistent style.
- we want to make it easy for them to do it the right way so that writing idiomatic code
- I see the ability to test & refactor as fairly important in working with and upgrading legacy codebases.
- We are not programming in the 90s anymore. 
and so so.

for me it smell like they want to change the basic further down the line, in favor of more c styled programming.
since I don't know the plans, I ask. maybe thats the idea, to make TB into that.

**Comment by wqweto** [comment]
Wayne wants TB to succeed. Everyone wants it. VB6 back compaitbility is a vehicle to TB being successful project.

The question one has to answer in this regard is *Are easy languages more successful than powerful languages*? You must admit that there are limits to what VB6 can do, there are areas where its BASIC nature prevents it from penetrating.

I'm not talking about writing drivers or operating systems. I'm talking about Windows game development. You're the expert -- tell me why is there no major studio, no major title **ever** made in VB6?

Another area VB6 is lacking strongly is professional mass-market desktop applications. Yes, the same LOBs most of VB6 developers toil day and night over.

When you look at the average Windows user Add/Remove Programs control panel chances are no more than 0.01% of the applications are VB6 made.

Guess what language are most of the applications (above 90%) there compiled with? What are 90% of the programs used in internet facing servers compiled with? (Hint: it's a letter, a letter and an increment operator)

**Reply by vbBaka** [reply]
well, yeah, I don't argue that (even if we could have a long debate about why VB6 never got into games)

what Im arguing is to improve VB6 by always placing VB6 first and if theres no way around it, try to find the closest possible method.
I don't understand why there should be limitations to use basic.
the reason my game is very fast is because I use Direct2D. even if the compiler is 20 years old, 32 bit and single-threaded.
so in the end its about that, to be able to use the machine the best way possible.
if Wayne is giving us multi-threading, 64bit, we of course need to adapt our coding to that.
but it should always be Basic-ish. I don't advocate for a stagnated TB, contrary, the next level basic, VB7 the next generation.
and I think u can achieve that in basic language.

but the OP question is to understand what is TwinBasic. really understand. and its purpose, the deep meaning, the far in the future goals.
where VB stands, to be the old that we need to forget or to renew VB into today and show it can do as well as any other languages.

**Reply by sharpninja** [reply]
> Another area VB6 is lacking strongly is professional mass-market desktop applications.

This won't happen without WinUI3 compatibility through the Win32 route.  Since WinUI is mostly winRT-style COM, that shouldn't be too difficult from a technical perspective, nut then you will have no support for GUI design using traditional VB methodology unless you create a set of WinUI controls to replicate the classic VB6 controls and a public api matching that of Win32.  What's the win for doing all of this?  I don't see it, honestly.  What I do see as a bright future for Twin Basic is allowing people to migrate VBA codebases to modern projects.

**Reply by mburns08109** [reply]
I think that the main reason that VB6 didn't penetrate the Gaming/"serious applications" marketplace was because of its need for the vb runtime dll (and the limits that it imposed).
With tB, that issue will now be moot.
With a decent, reasonably speedy compiler and no requirement for a massive framework like Java and .Net (granted, with a Win32-ish dependency that can be argued is comparable ~sorta) and with someone (or a group?) offering to help advance this tB toolset...? I'd say that the sky might be the limit - once word gets out in the "mainstream of programmer thinking" channels.

Think of this as an alternate programming ecosystem to the emerging "controls-based" methodology that the various frameworks are evolving towards at present. It will soon be very difficult to write "unapproved" code in those environments. tB will not suffer from those limitations (be that for good or ill). (I might add at this point - beware of the backlash that will follow for "breaking control". Expect it.)

**Comment by VB6-Programming** [comment]
@vbBaka 

I think your fears are unfounded. Wayne is doing everything possible to make twinBASIC as near to 100% backwards compatible with VB6 as possible (and making a brilliant job of it, too).
It's always been part of the plan to add new features, and obviously we are not all going to agree on what those new features should be or exactly how they should be implemented.
But they are **new** features, so none of us need use them if we don't wish to.

For me, the major new feature I want is 64 bit compiling - and we have got that :) and it is very impressive.
Obviously this means we need to be able to import existing VB6 apps (including forms). I have no doubt this will be achieved.  We can already import much VB6 code.

On the particular issue of operators:-
I'm not a fan of the  **A+=1**  style of operator (though I use it in other languages).  But Wayne is giving us both styles (and so maintaining compatibility).
For the bit shift operators, these weren't in VB6 so I voted  _NOT BOTHERED either way_  because I really don't see that it matters.

And to answer your question at a more 'conceptual' level, I would suggest that twinBASIC should become what VB6 would have been had Microsoft continued developing it.

**Reply by vbBaka** [reply]
the OP is not really about the shift operators, but more of understanding TB's spirit and what VB means, concretely.
I have the impression that some members, if able, they would start changing VB into something else. and eventually breaking compatibility, this to enable new ways to improve the old and crappy VB. while for me I want to keep that intact. 
I just want clarity.

**Reply by VB6-Programming** [reply]
I can't see Wayne allowing backwards compatibility to be broken (for Windows apps at least).

**Reply by bclothier** [reply]
Agreed. Nobody has advocated for breaking changes, only to make it better.

**Reply by mansellan** [reply]
I think that, once it has been decided that a new feature should be implemented, one of the most important next questions will be, "How can we add this feature in a way that is in harmony with the existing features (i.e. in a BASIC-y way)". I would hope that this gets enshrined into the [guiding principles](https://github.com/WaynePhillipsEA/twinbasic/discussions/610) when we get to that point.

I don't agree that people here hate VB6, on the contrary, we're here because we love it and want it to have a bright future. With regard to deprecation, I think that's just part of the lifecycle of any language still under active development. New features get added, which makes some particular piece of code easier, or more performant, or less likely to crash your app. At that point, it's nice if the IDE can give you a hint that there's now a new feature available that can help you out. But the original way is still there if you want it.

At work I'm a C# developer. If I want to use .Net features, I'd write some C#, not try and turn twinBASIC into C#. As @bclothier mentioned elsewhere, there's currently no native COM authoring language other than the horrid C / MFC / ATL franken-language. Yuck.

I got my start in IT though VB6, and still hold deep love for it. I certainly don't want to turn it into C (of any variant).

**Reply by FullValueRider** [reply]
My use of deprecated was not to indicate that something would be removed, but as a pointer that something better had been added to twinbasic.

**Comment by WaynePhillipsEA** [comment]
> @WaynePhillipsEA You are needed here because I don't understand something and I want clarity here. It seems that Im alone here, everything I say its going against. My understanding was the TwinBasic was the Continuation of VB6. it was born to create a modern version of VB6, and always have the fully compatibility in mind.

My main vision is quite simple.  My hope is that twinBASIC will evolve into what VB7 should have been.  In other words, it will be as if we were to imagine that development of VB6 had continued for the last 20 years instead of being abandoned in favour of VB.NET.  Backwards compatibility being of utmost importance, and with language features evolving in a considerate manner.

I have broader hopes for twinBASIC of course, such as replacing the VBIDE in Office... but that's later down the line.

> But, of what I can understand, people here seems to dislike VB6 and want, if they could, change "a lot" in favor for other syntaxes and methods.

Personally, I don't see this at all.  What I **_do_** see here is a lot of passion around the product (VB6/VBA -> VBx), and this is certainly great to see after all of these years where the product has stagnated.  For sure, ideas are being discussed openly... and there are some wacky ideas thrown around at times... but these discussions are exactly that:  **discussions** not **decisions**. 

> So what is the real purpose of TB? is it just to accommodate VB6 people, so they can "migrate" their project and when that is done its time to change?

Let me be clear: I've never made such a claim.  If you were to ask me; are there things that VB6 could do better?  Then my answer would be: of course!  No language is perfect, and that's why twinBASIC is here... to try to improve on the language, not change it beyond all recognition.

As @Kr00l rightly points out, all new features are **_optional_**.  You can continue programming in twinBASIC with the same fundamental language as VB6, ignoring all the new features, if that's what you really want.

> the spirit of keeping the VB6 "easiness" seems to be ignored. the response is "we need to evolve", and "all people want". like we are all experts. but the whole reason of Basic, is that to accommodate beginners and people that will never get to that expert level.
> and its not about one operator, that is nothing in comparison of what I can get from all the members here. theres no love to VB6 at all, just, old and wrongly made and now its time to fix it.

I believe you are wrong.  The reason people are here discussing twinBASIC so much is fundamentally because they love VBx!  Yes, some debates are lively, but everything I read suggests that we are **all** passionate about VBx, and want to see it succeed for another generation.  That's a great thing, and I hope it continues.

> so, thats the plan? to fix VB6 and turn it into a strange basic that people on expert level will enjoy using?

No, not at all!  We're trying to advance the language in ways that help facilitate powerful and more modern features, **_without_** affecting the ease and simpleness of the core language (which remains fully compatible with VBx).

**Reply by vbBaka** [reply]
well. that answered it all. thanks for tell all this.
also, I think its good to have this conversation so when someone new join, he/she can read this conversation if theres any doubts!
